ceno_zkvm/src/tables/program.rs

```diff
@@ -46,38 +49,56 @@ impl<T> InsnRecord<T> {
     pub fn as_slice(&self) -> &[T] {
         &self.0
     }
-
-    /// Iterate through the fields, except immediate because it is complicated.
-    fn without_imm(&self) -> &[T] {
-        &self.0[0..6]
-    }
-
-    /// The internal view of the immediate. See `DecodedInstruction::imm_internal`.
-    fn imm_internal(&self) -> &T {
-        &self.0[6]
-    }
 }
 
-impl InsnRecord<u32> {
+impl<F: SmallField> InsnRecord<F> {
     fn from_decoded(pc: u32, insn: &DecodedInstruction) -> Self {
         InsnRecord([
-            pc,
-            insn.opcode(),
-            insn.rd_internal(),
-            insn.funct3_or_zero(),
-            insn.rs1_or_zero(),
-            insn.rs2_or_zero(),
-            insn.imm_internal(),
+            (pc as u64).into(),
+            (insn.opcode() as u64).into(),
+            (insn.rd_internal() as u64).into(),
+            (insn.funct3_or_zero() as u64).into(),
+            (insn.rs1_or_zero() as u64).into(),
+            (insn.rs2_or_zero() as u64).into(),
+            Self::imm_internal_field(insn),
         ])
     }
 
     /// Interpret the immediate or funct7 as unsigned or signed depending on the instruction.
     /// Convert negative values from two's complement to field.
-    pub fn imm_internal_field<F: SmallField>(insn: &DecodedInstruction) -> F {
-        if insn.imm_field_is_negative() {
-            -F::from(-(insn.imm_internal() as i32) as u64)
+    pub fn imm_internal_field(insn: &DecodedInstruction) -> F {
+        let imm = Self::process_imm(insn);
+        if Self::imm_field_is_negative(insn) {
+            -F::from(-(imm as i32) as u64)
         } else {
-            F::from(insn.imm_internal() as u64)
+            F::from(imm as u64)
+        }
+    }
+
+    /// The internal view of the immediate, for use in circuits.
+    fn process_imm(insn: &DecodedInstruction) -> u32 {
+        let codes = insn.codes();
+        match codes.format {
+            R => codes.func7,
+            I if matches!(codes.kind, SLLI | SRLI | SRAI) => shift_imm::process_imm(insn),
+            _ => insn.immediate(),
+        }
+    }
+
+    /// The internal interpretation of the immediate sign, for use in circuits.
```

Comment:
Just report the immediate as a `i64` for use by circuits.  They can then directly encode that as a field element.

No need to make them (re-) do a bunch of logic.